semgrep-core/Parsing/other/yaml_to_generic.ml

```diff
@@ -0,0 +1,121 @@
+module Y = Yaml
+module S = Yaml.Stream
+module E = Yaml.Stream.Event
+module M = Yaml.Stream.Mark
+module A = AST_generic
+
+exception Unimplemented
+exception ParseError of string
+
+(* Parser for YAML. The tree-sitter grammar for YAML is somewhat complicated,
+ * since YAML is whitespace sensitive. Instead, we use the OCaml Yaml module,
+ * which we also use to parse YAML for semgrep. To get positions, we use the
+ * low-level Stream API, which we parse into a generic AST
+*)
+
+(* Helper functions *)
+
+let fst (a, _) = a
+
+let p_token = function
+  | E.Stream_start _ -> "stream start"
+  | E.Stream_end -> "stream end"
+  | E.Document_start _ -> "document start"
+  | E.Document_end _ -> "document end"
+  | E.Sequence_start _ -> "sequence start"
+  | E.Sequence_end -> "sequence end"
+  | E.Mapping_start _ -> "mapping start"
+  | E.Mapping_end -> "mapping end"
+  | E.Alias _ -> "alias"
+  | E.Scalar { value; _ } -> "scalar " ^ value
+  | E.Nothing -> "nothing"
+
+let mk_err err v {E.start_mark = {M.line; M.column; _}; _} =
+  Printf.sprintf "%s %s at line %d column %d" err (p_token v) line column
+
+let get_res = function
+  | Result.Error (`Msg s) -> raise (ParseError s)
+  | Result.Ok v -> v
+
+let any_of_list = function
+  | [x] -> A.E x
+  | xs -> A.E (A.Seq xs) (* Change to array? *)
+
+let tok (index, line, column) str file =
```

Comment:
You can look at the code of Parse_tree_sitter_helpers.token which also build token
from tree-sitter-info.
  (* Parse_info is 1-line based and 0-column based, like Emacs *)